paseos/activities/activity_manager.py

```diff
@@ -0,0 +1,144 @@
+                "Trying to add already existing activity with name: "
+                + name
+                + ". Already have "
+                + str(self._activities[name])
+            )
+
+        self._activities[name] = DotMap(
+            activity_function=activity_function,
+            power_consumption_in_watt=power_consumption_in_watt,
+            on_termination_function=on_termination_function,
+            constraint_function=constraint_function,
+            _dynamic=False,
+        )
+
+        logger.debug(f"Registered activity {self._activities[name]}")
+
+    def perform_activity(
+        self,
+        name: str,
+        activity_func_args: list = None,
+        termination_func_args: list = None,
+        constraint_func_args: list = None,
+    ):
+        """Perform the specified activity. Will advance the simulation if automatic clock is not disabled.
+
+        Args:
+            name (str): Name of the activity to perform.
+            activity_func_args (list, optional): Arguments for the activity function. Defaults to None.
+            termination_func_args (list, optional): Arguments for the termination function. Defaults to None.
+            constraint_func_args (list, optional): Arguments for the constraint function. Defaults to None.
+        """
+        # Check if activity exists and if it already had consumption specified
+        assert (
+            name in self._activities.keys()
+        ), f"Activity not found. Declared activities are {self._activities.keys()}"
+        activity = self._activities[name]
+        logger.debug(f"Performing activity {activity}")
+
+        assert (
+            activity.power_consumption_in_watt >= 0
+        ), "Power consumption has to be positive but was specified as " + str(
+            activity.power_consumption_in_watt
+        )
+
+        activity_runner = ActivityRunner(
+            name=name,
+            activity_func=activity.activity_function,
+            constraint_func=activity.constraint_function,
+            termination_func=activity.on_termination_function,
+            termination_args=termination_func_args,
+            constraint_args=constraint_func_args,
+        )
+
+        async def job():
+            processor = ActivityProcessor(
+                update_interval=self._paseos_update_interval,
+                power_consumption_in_watt=activity.power_consumption_in_watt,
+                paseos_instance=self._paseos_instance,
+                activity_runner=activity_runner,
+                advance_paseos_clock=self._paseos_instance.use_automatic_clock,
+            )
+
+            await asyncio.wait(
+                [
+                    processor.start(),  # noqa: F821
+                    activity_runner.start(activity_func_args),
+                ],
+                return_when=asyncio.ALL_COMPLETED,
+            )
+            self._paseos_instance._is_running_activity = False
+            del processor
```

Comment:
Do we need to delete the processor every time? Would it make sense to create an ActivityProcessor in the ActivityManager constructor and simply update the ActivityProcessor when perform_activity()?

Reply:
This was an intentional design choice. If for some reason in the future we have different ways of performing activities or interact with the processor in another way, it is safest, I thought, if the lifetime of the singleton processor is exactly the time performing the activity and if an exception is thrown if you try to create another one while the activity is running. It seems also less error prone this way, I think, because if you e.g. forget to update one of the functions or parameters of the old processor properly you might execute the wrong code without noticing?